Fix grpc+gdr compile error introduced in 33170cc6.

byronyi opened the pull request:
@poxvoculi Most CPU tensors in PS are still managed by cpu_allocator, and I have added a separate path to copy them back to a CPU tensor managed by ProcessState::GetCPUAllocator.

I believe there is no harm (maybe a little bit larger page table on NIC?) to register the same memory region multiple times, so it is not required for the memory managers to be singletons.

poxvoculi commented:
@byronyi Sorry for breaking this directory.  I don't have an environment in which I can compile and test it, so I ended up submitting a best guess.

With the latest change, the value returned by cpu_allocator() should not be directly passed to any devices if ProcessState::AddCPUAllocVisitor is called prior to any CPU devices being created.   It's important when RDMA is going to be used that visitors be registered very early in process startup, so that all uses get the correct allocators.    I tried to achieve that by putting the new RegMemVisitors in the GdrMemoryManager constructor, which gets called in the GdrServer constructor.  The devices are not created until GrpcServer::Init, [here](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/distributed_runtime/rpc/grpc_server_lib.cc#L162).  Are you seeing the PS processes using the cpu_allocator() values directly, and not the allocator with SubAllocator using the registered visitors, created [here?](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/common_runtime/process_state.cc#L91)

Whether it's a problem to duplicate-register the same region depends on the networking libraries in use.  For us it's a fatal error, but I can imagine another library ignoring a duplicate registration or making a duplicate table entry.

byronyi commented:
It's quite alright. In fact, with Ubuntu 18.04 or any Linux distro with kernel > 4.14, we could build the package without using any specific hardware or proprietary drivers/libraries, like in a GCP VM or a plain container. Testing is also possible with SoftRoCE (with in-tree driver rdma-rxe). I will add detailed docs after we move the current code to tensorflow/networking.

As to the tensors allocated by cpu_allocator(), I found many of them are 1-byte boolean flag, like `report_uninitialized_variables/boolean_mask/Prod`. You could take a look of the detailed log [here](https://gist.github.com/byronyi/568998c100615dfc9cada5112c7fd05c) and [there](https://gist.github.com/byronyi/e4ba356dfb40c135dc673acc36b9424f).

byronyi commented:
By the way, should we use 0 or -1 for unknown NUMA node? I would like to stay consistent to the rest of core.

poxvoculi commented:
Thanks for the logs.  What's going on with the direct use of cpu_allocator is that there are still a few uses in the code base that haven't been cleaned up yet, for example in Tensor::FromProto.  I'll be fixing them soonish, but it may be some weeks before they're all taken care of, so if you can handle doing duplicate registration for now, that's probably best.

For unknown NUMA node, you should be using port::kNUMANoAffinity, defined in tensorflow/core/platform/numa.h.  For the present time this will continue to resolve to CPU:0.  In the future it will indicate a "don't care" or "no particular affinity" choice.

byronyi commented:
@poxvoculi Hi Paul, I rebased to current master. Let me know if you have any comments. Thanks!

byronyi committed:
Added log entries for copying unpinned memory RDMA

Currently there are large number of tensors managed
by non-visitable memory allocators in CPU-only PS.
GPU workers seem less prone to this problem.

Copying large sized tensor buffers may introduce
non-trivial overhead. Should probably fix this.

Signed-off-by: Bairen Yi <byi@connect.ust.hk>

byronyi commented:
@caisq Is this ready to pull?

byronyi commented:
@poxvoculi Hi Paul, is this ready to pull?

poxvoculi commented:
Yes, I think it's ready.  I'll try to find out who's supposed to be doing this.

caisq commented:
@byronyi The PR will be merged automatically when it passes tests and gets submitted internally.

yanivbl6 commented:
hi @byronyi, @poxvoculi 
I tried running the new GDR from latest master but I encountered an error:
Check failed: 0 == cpu_allocators_.size() (0 vs. 1)AddCPUAllocVisitor must be called prior to first call to ProcessState::GetCPUAllocator

I am also getting the same error on my the patch I am working on to fix the verbs contribution.

My understanding is the the GPU "base device" is created in [this line](https://github.com/tensorflow/tensorflow/blob/aab3c53e1484404a70565324d1231c4e6ead7425/tensorflow/core/common_runtime/gpu/gpu_device.cc#L1111), trigerring the call to GetCPUAllocator before the visitors are registered, which is an error.
I am still trying to figure out how the tensorflow core is managing/syncing those event (GPU context creation and memory manager initialization).

byronyi commented:
@yanivbl6 Hi Yaniv, could you elaborate on your running environment? Is it PS/worker, and is GPU involved?

poxvoculi commented:
This could be an affect of the organization of your program at the python level.  Because variables have a lifespan that outlasts sessions, allocator initialization happens once per process.  It's important that all session creation in the python expressed TF program be consistent, i.e. if the program is going to run distributed then don't ever create a local-only session.  If you create a local-only session then later a distributed session, something like this error might occur.

yanivbl6 commented:
I don't think it's coming from the program, since I was running the upstream tf_cnn benchmark (synthetic run), and tried reverting back to older versions (back till tf1.5 compatible), with the same error.

Also- the error is happening on initialization, so the variables don't yet have a session to outlast.
I will try investigating the issue a bit further and will open an issue once I know more.

byronyi commented:
@yanivbl6 I could reproduce that on a DGX-1V. Will take a few days to fix it.

yanivbl6 commented:
thanks for checking it- I am also working on it and I will make sure to patch GDR if I find a solution.
